internal/gengapic/genrest_test.go

```diff
@@ -0,0 +1,231 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gengapic
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/protobuf/protoc-gen-go/descriptor"
+	plugin "github.com/golang/protobuf/protoc-gen-go/plugin"
+	"github.com/googleapis/gapic-generator-go/internal/pbinfo"
+	"google.golang.org/genproto/googleapis/api/annotations"
+)
+
+// This is a hack to provide an lvalue for a constant.
+func typep(typ descriptor.FieldDescriptorProto_Type) *descriptor.FieldDescriptorProto_Type {
+	return &typ
+}
+
+// Another hack, this time for field numbers.
+func idxAddr(i int) *int32 {
+	j := int32(i)
+	return &j
+}
+
+func setupMethod(t *testing.T, g *generator, url string, allFieldNames []string) *descriptor.MethodDescriptorProto {
+	msg := &descriptor.DescriptorProto{
+		Name: proto.String("IdentifyRequest"),
+	}
+	for i, name := range allFieldNames {
+		j := int32(i)
+		msg.Field = append(msg.Field,
+			&descriptor.FieldDescriptorProto{
+				Name:   proto.String(name),
+				Number: &j,
+				Type:   typep(descriptor.FieldDescriptorProto_TYPE_INT32),
+			},
+		)
+	}
+	mthd := &descriptor.MethodDescriptorProto{
+		Name:      proto.String("Identify"),
+		InputType: proto.String(".identify.IdentifyRequest"),
+		Options:   &descriptor.MethodOptions{},
+	}
+
+	// Just use Get for everything and assume parsing general verbs is tested elsewhere.
+	if err := proto.SetExtension(mthd.Options, annotations.E_Http, &annotations.HttpRule{
+		Pattern: &annotations.HttpRule_Get{
+			Get: url,
+		},
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	srv := &descriptor.ServiceDescriptorProto{
+		Name:    proto.String("IdentifyMolluscService"),
+		Method:  []*descriptor.MethodDescriptorProto{mthd},
+		Options: &descriptor.ServiceOptions{},
+	}
+	if err := proto.SetExtension(srv.Options, annotations.E_DefaultHost, proto.String("linnaean.taxonomy.com")); err != nil {
+		t.Fatal(err)
+	}
+
+	fds := []*descriptor.FileDescriptorProto{
+		&descriptor.FileDescriptorProto{
+			Package:     proto.String("identify"),
+			Service:     []*descriptor.ServiceDescriptorProto{srv},
+			MessageType: []*descriptor.DescriptorProto{msg},
+		},
+	}
+	req := plugin.CodeGeneratorRequest{
+		Parameter: proto.String("go-gapic-package=path;mypackage,transport=rest"),
+		ProtoFile: fds,
+	}
+	g.init(&req)
+
+	return mthd
+}
+
+func TestPathParams(t *testing.T) {
+	var g generator
+	g.apiName = "Awesome Mollusc API"
+	g.imports = map[pbinfo.ImportSpec]bool{}
+	g.opts = &options{transports: []transport{rest}}
+
+	for tstNum, tst := range []struct {
+		url           string
+		allFieldNames []string
+		expected      map[string]*descriptor.FieldDescriptorProto
+	}{
+		// Null case: no params
```

Comment:
Nit, suggestion: add a `name` field to the test case and put this mini description there, or something similar, then in `t.Errorf` you can report the exact test that failed rather reporting the index of the failed case, then having to count manually :)